### PR TITLE
feat: Fisher shrinkage + exact (true) Fisher accumulation

### PR DESCRIPTION
**Pull request by hmousset**
# feat: Fisher shrinkage + exact (true) Fisher accumulation

## Summary

Two additions to the `use_fisher` preconditioning path:

1. `fisher_shrinkage`: ridge-shrinks the gradient-covariance matrix `E` so whitening doesn't collapse to near-zero rank.
2. Exact ("true") Fisher accumulation for classification: replaces the empirical-Fisher approximation with the analytic softmax Fisher.

## Motivation

`use_fisher=True` whitens `E` with an absolute eigenvalue cutoff (`1e-6`): any eigenvalue below it is dropped before inverting. At step 0, `E`'s whole spectrum can sit near that cutoff, so this truncation collapses `E` to a very small rank instead of its real rank and with only a couple of directions left, the resulting scores go winner-take-all across layers. I suspect this is one of the mechanisms behind the U-shaped depth allocation seen in GrowRA experiments
Separately, we've only ever used the *empirical* Fisher (outer products of the real-label gradient) as a stand-in for the true Fisher. For a softmax classifier the true Fisher of the logits is analytic, `F_z = diag(p) - p p^T` (rank `K-1`), and exact `E` can be obtained by backpropagating its `K-1` scaled eigenvectors as gradient seeds instead of the one real-label gradient.

The two are independent: shrinkage regularizes whichever `E` you have (true or "empirical"), true-Fisher accumulation changes how `E` is built.

### `fisher_shrinkage`

```python
model.compute_optimal_updates(use_fisher=True, fisher_shrinkage=0.1)  
```

- Only matters when `use_fisher=True`. `0.0` (default) = old behavior, bit-exact. `1.0` = `E` replaced by a scaled identity.
- Wired through `compute_optimal_delta`, `compute_optimal_updates` (linear + conv2d), `GrowingBlock.compute_optimal_updates`. Not supported on `GrowingDAG` (`use_fisher=True` now raises `NotImplementedError` there — DAG delta goes through `MergeGrowingModule`, which has no `E`).

### True Fisher accumulation (classification only)

```python
logits = model(x)
loss = criterion(logits, labels)
model.update_computation_true_fisher(loss, logits)   # replaces loss.backward() + update_computation()
...
model.compute_optimal_updates(use_fisher=True, fisher_shrinkage=0.1)  # E only gets read if use_fisher=True
```

Manual two-step form, if you need the real-label backward yourself first:

```python
loss.backward(retain_graph=True)
model.update_computation(update_covariance_loss_gradient=False)  # S, M only
model.accumulate_true_fisher_covariance(logits)                  # E, exact
```

`update_computation_true_fisher`/`accumulate_true_fisher_covariance` only control how `E` is *written*; `use_fisher` still controls whether it's *read*. Accumulating the true Fisher and then calling `compute_optimal_updates(use_fisher=False)` computes exact `E` for nothing — it's never consulted. `use_fisher=True` still applies to the plain empirical `E`.

Notes:
- `logits`: `(batch, n_classes)`, pre-softmax, same forward pass as `loss`, still attached to the graph.
- Costs `K-1` extra backward passes (1 for binary).
- `loss_normalization` defaults to batch size (mean-reduced cross-entropy); pass your own if different so `S`/`M` and `E` share a scale.
- 
## Testing

New tests in `test_linear_growing_module.py`, `test_tools.py`, `test_growing_dag.py`, `test_growing_graph_network.py`: closed-form shrinkage vs. SVD ground truth, shrinkage=0 no-op, out-of-range rejection, truncated-`E` rescue, scale invariance, `count_samples` semantics, `clear_pre_activity_grad` incl. merge-module delegation, true-Fisher closed-form match for K=2,3, binary = 1 backward, `NotImplementedError` on `GrowingDAG`, dispatch through `GrowingGraphNetwork`.

### CODE REVIEW

File: tests/test_tools.py [hunk 1]
```diff
@@ -397,6 +397,45 @@ def test_compute_optimal_added_parameters(self):
         )
         self.assertTrue(torch.allclose(omega, torch.zeros_like(omega)))
 
+    def test_compute_optimal_added_parameters_e_numerical_threshold(self):
+        """E whitening cutoff can differ from S's via e_numerical_threshold."""
+        torch.manual_seed(42)
+        matrix_s = torch.eye(3) * 2.0
```
Comment (TheoRudkiewicz) [inline tests/test_tools.py:402]:
```suggestion
    def test_compute_optimal_added_parameters_e_numerical_threshold(self):
        """E whitening cutoff can differ from S's via e_numerical_threshold."""
        matrix_s = torch.eye(3) * 2.0
```
Seed is already set in the `TorchTestCase.setUp`

File: src/gromo/modules/growing_module.py [hunk 2]
```diff
@@ -22,6 +22,38 @@
...
+        the shrunk matrix E
+    """
+    assert 0.0 <= fisher_shrinkage <= 1.0, (
+        f"fisher_shrinkage must be in [0, 1], got {fisher_shrinkage}"
+    )
+    d = matrix_e.shape[0]
+    return (1 - fisher_shrinkage) * matrix_e + fisher_shrinkage * (
+        matrix_e.trace() / d
+    ) * torch.eye(d, device=matrix_e.device, dtype=matrix_e.dtype)
+
```
Comment (TheoRudkiewicz) [inline src/gromo/modules/growing_module.py:49]: Maybe this could documented as generic i.e. not only for empirical Fisher but for any SP(D) matrix.

**Review by TheoRudkiewicz — APPROVED** [review]
This seems good to me. I added minor comments.

File: src/gromo/modules/growing_module.py (hunk 2 above)
Comment (TheoRudkiewicz) [reply src/gromo/modules/growing_module.py:49]: Done

**Review by sylvchev — APPROVED** [review]
Thanks @hmousset and @TheoRudkiewicz 
Let get this merge!
